gamenet/teeworlds-0.7/src/msg/game.rs

```diff
@@ -1486,17 +1491,33 @@ impl<'a> fmt::Debug for SvClientDrop<'a> {
 
 impl SvGameMsg {
     pub fn decode<W: Warn<Warning>>(warn: &mut W, _p: &mut Unpacker) -> Result<SvGameMsg, Error> {
-        let result = Ok(SvGameMsg);
+        let result = Ok(SvGameMsg {
+            game_msg: enums::Gamemsg::from_i32(_p.read_int(warn)?)?,
+            para_I: _p.read_int(warn).ok(),
+            para_II: _p.read_int(warn).ok(),
+            para_III: _p.read_int(warn).ok(),
+        });
         _p.finish(warn);
         result
     }
     pub fn encode<'d, 's>(&self, mut _p: Packer<'d, 's>) -> Result<&'d [u8], CapacityError> {
+        assert!(self.para_I.is_some());
+        assert!(self.para_II.is_some());
+        assert!(self.para_III.is_some());
```

Comment:
I would consider sending some default like 0 instead of nothing wrong in that case. Those parameters are not always sent. Unused defaults seem misleading at best and break the protocol at worst. Doesn't rust have an optional type? How is the default supposed to be represented in the json?

Reply:
> I would consider sending some default like 0 instead of nothing wrong in that case. Those parameters are not always sent. Unused defaults seem misleading at best and break the protocol at worst.

I tend to agree.

> Doesn't rust have an optional type?

Yes, it does.

>  How is the default supposed to be represented in the json?

Currently, the JSON description does not support optional fields. It only supports fields that can be omitted for backwards compatibility. If we want to accurately support this 0.7 message, we'd first have to figure out a way to represent this in the JSON and then adapt all users of this JSON to deal with it.

Reply:
> we'd first have to figure out a way to represent this in the JSON

Yes that is what I am asking about. I will not come up with the format if you have your ideas anyways. I have no good idea so that one is on you.

> and then adapt all users of this JSON to deal with it.

By users do you mean humans with their libtw2 based projects? Or parts of the code in libtw2 that consume the json? Either way do you have a list of those users?

Reply:
Perhaps we could specify that this is a repetitions of i32s to the end of the message. That sounds kinda sane to me. Perhaps `"kind":"array-until-end"`?

Parts of libtw2 that consume the JSON, wireshark dissector and code generation for the gamenet crates, I think.

Reply:
Yes I like the idea sounds good. I sadly already spent way more time on this pr than initially planned. And it turned into something that got way bigger than I have motivation to work on.

I am closing this pr now so it does not get stale.

I would appreciate if one day you add support for the sv game msg to the dissector. Until then I will just use my branch that added this message because it works wonderfully for my use case which is debugging traffic including this very game message.